### PR TITLE
Change to enable full iteration CUDA graph for LLMs

### PR DESCRIPTION
**Pull request by vasunvidia**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Enables Full iteration CUDA graph for NeMo models.

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/diffusion/models/dit/dit_layer_spec.py [hunk 1]
```diff
@@ -579,7 +579,7 @@ def __init__(
         hidden_size = config.hidden_size
         super().__init__(config=config, submodules=submodules, layer_number=layer_number)
 
-        if config.enable_cuda_graph:
+        if config.enable_cuda_graph and config.cuda_graph_scope != "full_iteration":
```
Comment (katec846) [inline nemo/collections/diffusion/models/dit/dit_layer_spec.py:583]: I'm quite confused about how the flags should be set. Please correct me if I'm wrong

- To enable partial CUDA graph, we should set `config.enable_cuda_graph` to True (I believe `config.cuda_graph_scope` defaults to "full", which captures a whole Transformer layer—though I find this flag a bit confusing).
- To enable full-iteration CUDA graph, should we set `enable_full_iteration_cuda_graph` to True? But I don't see this flag introduced anywhere. Wouldn't it be better if, when we set `config.enable_cuda_graph` to True and `config.cuda_graph_scope='full_iteration'`, the code automatically set `enable_full_iteration_cuda_graph` to true somewhere?

File: nemo/lightning/megatron_parallel.py [hunk 2]
```diff
@@ -160,6 +161,7 @@ class MegatronParallel(nn.ModuleList, Generic[ModelT]):
         cpu (bool): Whether model should reside on CPU.
         convert_module_fn (Optional[Callable[[ModelT], nn.Module]]): An optional function to
             apply to the model parameters after initialization.
+        enable_full_iteration_cuda_graph (bool): Whether Full iteration CUDAgraph should be enabled.
```
Comment (katec846) [inline nemo/lightning/megatron_parallel.py]: I believe in your Megatron-LM MR, you included this

>     if args.enable_cuda_graph and args.cuda_graph_scope=="full_iteration":
>         assert not args.check_for_nan_in_loss_and_grad, \
>             "--no-check-for-nan-in-loss-and-grad should be set with full_iteration CUDA graph"

We also need this comment and assertion in Nemo

File: nemo/collections/diffusion/models/dit/dit_layer_spec.py (hunk 1 above)
Comment (vasunvidia) [reply nemo/collections/diffusion/models/dit/dit_layer_spec.py:583]: Regarding config.cuda_graph_scope defaults to "full", this will be removed in MCore MR https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/3664
Thanks for the comment to use config.enable_cuda_graph for "full_iteration" CUDA graph. I added this change and this simplifies the change quite a bit.

Comment (erhoo82) [inline nemo/collections/diffusion/models/dit/dit_layer_spec.py:583]: Please a comment of `Enable per-Transformer layer cuda graph` above

File: nemo/collections/diffusion/models/dit/dit_layer_spec.py [hunk 3]
```diff
@@ -695,7 +695,7 @@ def __init__(
     ):
         super().__init__(config=config, submodules=submodules, layer_number=layer_number)
 
-        if config.enable_cuda_graph:
+        if config.enable_cuda_graph and config.cuda_graph_scope != "full_iteration":
```
Comment (erhoo82) [inline nemo/collections/diffusion/models/dit/dit_layer_spec.py:700]: Please a comment of `Enable per-Transformer layer cuda graph` above

File: nemo/collections/llm/gpt/model/base.py [hunk 4]
```diff
@@ -326,7 +326,7 @@ def configure_model(self, tokenizer, pre_process=None, post_process=None, vp_sta
         Returns:
             MCoreGPTModel: Configured Megatron Core GPT model instance
         """
-        if self.enable_cuda_graph:
+        if self.enable_cuda_graph and self.cuda_graph_scope != "full_iteration":
```
Comment (erhoo82) [inline nemo/collections/llm/gpt/model/base.py:330]: Please a comment of `Enable per-Transformer layer cuda graph` above

File: nemo/lightning/megatron_parallel.py [hunk 5]
```diff
@@ -583,7 +584,8 @@ def init_model_parallel(self):
         # Skip init_ddp for inference i.e testing as it can lead to OOM.
         try:
             if not self.trainer.state.fn == TrainerFn.TESTING:
-                self.init_ddp()
+                with torch.cuda.stream(torch.cuda.Stream()):
```
Comment (erhoo82) [inline nemo/lightning/megatron_parallel.py:598]: Please comment why we do init_dpp under a side stream

File: nemo/lightning/megatron_parallel.py (hunk 2 above)
Comment (vasunvidia) [reply nemo/lightning/megatron_parallel.py]: For this, we need to add a check in MCore in a separate MR.

File: nemo/collections/diffusion/models/dit/dit_layer_spec.py (hunk 1 above)
Comment (vasunvidia) [reply nemo/collections/diffusion/models/dit/dit_layer_spec.py:583]: Done.

File: nemo/collections/llm/gpt/model/base.py (hunk 4 above)
Comment (vasunvidia) [reply nemo/collections/llm/gpt/model/base.py:330]: Done

File: nemo/lightning/megatron_parallel.py (hunk 5 above)
Comment (vasunvidia) [reply nemo/lightning/megatron_parallel.py:598]: Done.

File: nemo/collections/diffusion/models/dit/dit_layer_spec.py (hunk 3 above)
Comment (vasunvidia) [reply nemo/collections/diffusion/models/dit/dit_layer_spec.py:700]: Done

**Review by katec846 — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

File: scripts/performance/helpers.py [hunk 6]
```diff
@@ -316,6 +374,9 @@ def set_perf_optimization_configs(
         enable_cuda_graphs = False
     recipe = set_cuda_graph_configs(recipe, enable_cuda_graphs, task)
 
+    if enable_cuda_graphs:
```
Comment (vasunvidia) [inline scripts/performance/helpers.py:377]: We also need to check for cuda_graph_scope=="full_iteration" here?

Comment (vasunvidia) [reply scripts/performance/helpers.py:377]: Oh. I see the check in set_full_iteration_cuda_graph_configs.
